### ISSUE TITLE
Annotations/label filtering on on.delete() handler doesn't seem to work

### ISSUE THREAD

**Issue by olivier-mauras** [zalando-incubator/kopf#167]
## Expected Behavior
With an on.delete() handler filtering on annotations, I expected to retrieve the kopf finalizer only on filtered resources.

## Actual Behavior
All resources handled by the on.delete() handler end up with the kopf finalizer making me believe the filtering either by annotation or label to not work correctly

## Steps to Reproduce the Problem
``` python
@kopf.on.delete('', 'v1', 'serviceaccounts', annotations={'custom/created_by': 'namespace-manager'})
async def sa_delete(body, namespace,  **kwargs):
  pass
```

Result:
``` text
$ kubectl -n trident get sa tiller -o yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  annotations:
    kopf.zalando.org/last-handled-configuration: '{"secrets": [{"name": "tiller-token-mxn8p"}],
      "metadata": {"labels": {"custom/name":      "trident", "custom/team": "trident"}, "annotations": {"custom/created_by":
      "namespace-manager"}}}'
    custom/created_by: namespace-manager
  creationTimestamp: 2019-08-06T06:17:51Z
  finalizers:
  - kopf.zalando.org/KopfFinalizerMarker
  labels:
    custom/name: trident
    custom/team: trident
  name: tiller
  namespace: trident
secrets:
- name: tiller-token-mxn8p

$ kubectl -n trident get sa default -o yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  annotations:
    kopf.zalando.org/last-handled-configuration: '{"secrets": [{"name": "default-token-bzjqv"}]}'
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","kind":"ServiceAccount","metadata":{"annotations":{"kopf.zalando.org/last-handled-configuration":"{\"secrets\": [{\"name\": \"default-token-bzjqv\"}]}"},"creationTimestamp":"2019-02-04T12:18:54Z","name":"default","namespace":"trident","resourceVersion":"27507246","selfLink":"/api/v1/namespaces/trident/serviceaccounts/default","uid":"0a321e2f-2877-11e9-a525-005056abd413"},"secrets":[{"name":"default-token-bzjqv"}]}
  creationTimestamp: 2019-02-04T12:18:54Z
  finalizers:
  - kopf.zalando.org/KopfFinalizerMarker
  name: default
  namespace: trident
secrets:
- name: default-token-bzjqv
```

By my understanding only `tiller` service account should have been matched and handled but sadly `default` service account gets the finalizer as well.
Did I miss something??

**Comment by dlmiddlecote** [comment]
Hey!

I actually contributed in both of these areas, finalizers and filtering by annotations, and I can confirm that this is a bug. The finalizer addition is ignoring the fact that the object doesn't trigger any handlers.

I've added a PR for this (will add tests shortly), but I guess it's up to the maintainer @nolar to confirm + merge this.

Sorry for the bother!
Dan.

**Comment by nolar** [comment]
@dlmiddlecote The bug is confirmed by @psycho-ir. I would appreciate if you fix it (and add the unit-tests for this case), while I'm trying to clean up a pile of unrelated PRs and branches on my side.

**Comment by nolar** [comment]
The fix is pre-released as [kopf==0.21rc3](https://github.com/zalando-incubator/kopf/releases/tag/0.21rc3).

But beware of other massive changes in rc1+rc2+rc3 combined (see [Releases](https://github.com/zalando-incubator/kopf/releases)). Other things can break. We will test it in real action before releasing 0.21.

**Comment by nolar** [comment]
It is now in [`kopf==0.21`](https://github.com/zalando-incubator/kopf/releases/tag/0.21). If the issue reappears, feel free to re-open.
